pkg/tunnels/linux_tunnels.go

```diff
@@ -0,0 +1,368 @@
+package tunnels
+
+import (
+	"bufio"
+	"crypto/sha256"
+	"fmt"
+	"net"
+	"os/exec"
+	"strconv"
+	"strings"
+
+	"github.com/cloudnativelabs/kube-router/v2/pkg/routes"
+	"github.com/cloudnativelabs/kube-router/v2/pkg/utils"
+	"github.com/vishvananda/netlink"
+	"k8s.io/klog/v2"
+)
+
+const (
+	EncapTypeFOU  = EncapType("fou")
+	EncapTypeIPIP = EncapType("ipip")
+
+	// FOU modes used for the iproute2 tooling
+	fouIPv4LinkMode = "ipip"
+	fouIPv6LinkMode = "ip6tnl"
+
+	// IPIP modes used for the iproute2 tooling
+	ipipIPv4Mode = "ipip"
+	ipipIPv6Mode = "ip6ip6"
+
+	// The maximum and minimum port numbers for encap ports
+	maxPort = uint16(65535)
+	minPort = uint16(1024)
+)
+
+var (
+	validEncapTypes = []EncapType{EncapTypeFOU, EncapTypeIPIP}
+)
+
+// EncapType represents the type of encapsulation used for an overlay tunnel in kube-router.
+type EncapType string
+
+// IsValid checks if the encapsulation type is valid by comparing it against a list of valid types.
+// It returns true if the encapsulation type is valid, otherwise it returns false.
+func (e EncapType) IsValid() bool {
+	for _, validType := range validEncapTypes {
+		if e == validType {
+			return true
+		}
+	}
+	return false
+}
+
+// ParseEncapType parses the given string and returns an Encap type if valid.
+// It returns an error if the encapsulation type is invalid.
+//
+// Parameters:
+//   - s: A string representing the encapsulation type.
+//
+// Returns:
+//   - Encap: The parsed encapsulation type.
+//   - error: An error if the encapsulation type is invalid.
+func ParseEncapType(encapType string) (EncapType, error) {
+	encap := EncapType(encapType)
+	if !encap.IsValid() {
+		return "", fmt.Errorf("invalid encapsulation type: %s", encapType)
+	}
+	return encap, nil
+}
+
+type EncapPort uint16
+
+func (e EncapPort) checkWithinRange() error {
+	if uint16(e) > minPort && uint16(e) < maxPort {
+		return nil
+	}
+	return fmt.Errorf("specified encap port is out of range of valid ports: %d, valid range is from %d to %d",
+		e, minPort, maxPort)
+}
+
+func ParseEncapPort(encapPort uint16) (EncapPort, error) {
+	port := EncapPort(encapPort)
+	if err := port.checkWithinRange(); err != nil {
+		return 0, err
+	}
+	return port, nil
+}
+
+type Tunneler interface {
+	SetupOverlayTunnel(tunnelName string, nextHop net.IP, nextHopSubnet *net.IPNet) (netlink.Link, error)
+	EncapType() EncapType
+	EncapPort() EncapPort
+}
+
+type OverlayTunnel struct {
+	krNode           utils.NodeIPAndFamilyAware
+	overlayEncapPort EncapPort
+	overlayEncap     EncapType
+}
+
+func NewOverlayTunnel(krNode utils.NodeIPAndFamilyAware, overlayEncap EncapType,
```

Comment:
This might be a good one to have a conversation about. Coming from a different programming background than GoLang, I still have the tendency to not make fields public any time that I don't have to. I guess I have been considering it a forcing function for using functions & interfaces rather than getting used to changing the data on a struct directly.

Not sure if this is a good practice in Go or not though. Do you have any thoughts surrounding this?

Reply:
I used to have the same tendencies, but I more and more accepted that Go is the way Go is, and that it's better to keep it as straight forward as possible. Certain things can't be avoided anyways (like constructing zero values), so callers will almost always be able to break certain invariants that you try to uphold in your types.

The main benefit of having this function that I can see is that you will get compiler errors at the call sites when new required fields are added to `OverlayType`, so you can't miss to update all of them. If that's important, then it might be a good idea to keep it. On the other hand, adding optional fields is harder with this approach, but darn easy when you have a struct with public fields. I tend to think of those structs as "factory functions with named parameters".

The amount of work you have to put into your types to make them hard to misuse is (subjectively) higher in Go than in some other popular languages out there. This is a trade-off. Do you have a type that will be used all over the place, or is it something that will only be used a few times, maybe just once? Putting more effort (i.e. implementing patterns that require quite a few more lines of code) into the former will pay off more in the long run than for the latter.

Reply:
Thanks for taking the time to write out how you think about coding in Go! :smile: 

In this case, I think that I'll keep this particular one, but in general, I think that I agree with you. It does seem that Go is meant to be simpler and more straight forward. Trying to hide everything inside a struct that is only used rarely, is probably not worth the effort and is likely antithetical to the design of Go.

I'll keep this in mind during the next couple of PRs and see how it works for me.